Patch vtable slots and similar when tiering is enabled

kouvel opened the pull request:
For a method eligible for tiered compilation and vtable slot backpatching:
  - The entry point to the final code is versionable, as for any method eligible for tiering
  - It does not have a precode (`HasPrecode()` returns false)
  - It does not have a stable entry point (`HasStableEntryPoint()` returns false)
  - A call to the method may be:
    - An indirect call through the `MethodTable`'s backpatchable vtable slot
    - A direct call to a backpatchable `FuncPtrStub`, perhaps through a `JumpStub`
    - For interface methods, an indirect call through the virtual stub dispatch (VSD) indirection cell to a backpatchable `DispatchStub` or a `ResolveStub` that refers to a backpatchable `ResolveCacheEntry`
  - The purpose is that typical calls to the method have no additional overhead when tiering is enabled

Recording and backpatching slots:
  - In order for all vtable slots for the method to be backpatchable:
    - A vtable slot initially points to the `MethodDesc`'s temporary entry point, even when the method is inherited by a derived type (the slot's value is not copied from the parent)
    - The temporary entry point always points to the prestub and is never backpatched, in order to be able to discover new vtable slots through which the method may be called
    - The prestub, as part of `DoBackpatch()`, records any slots that are transitioned from the temporary entry point to the method's at-the-time current, non-prestub entry point
    - Any further changes to the method's entry point cause recorded slots to be backpatched in `BackpatchEntryPointSlots()`
  - In order for the `FuncPtrStub` to be backpatchable:
    - After the `FuncPtrStub` is created and exposed, it is patched to point to the method's at-the-time current entry point if necessary
    - Any further changes to the method's entry point cause the `FuncPtrStub` to be backpatched in `BackpatchEntryPointSlots()`
  - In order for VSD entities to be backpatchable:
    - A `DispatchStub`'s entry point target is aligned and recorded for backpatching in `BackpatchEntryPointSlots()`
    - A `ResolveCacheEntry`'s entry point target is recorded for backpatching in `BackpatchEntryPointSlots()`

Slot lifetime and management of recorded slots:
  - A slot is recorded in the `LoaderAllocator` in which the slot is allocated, see `RecordAndBackpatchEntryPointSlot()`
  - An inherited slot that has a shorter lifetime than the `MethodDesc`, when recorded, needs to be accessible by the `MethodDesc` for backpatching, so the dependent `LoaderAllocator` with the slot to backpatch is also recorded in the `MethodDesc`'s `LoaderAllocator`, see `MethodDescBackpatchInfo::AddDependentLoaderAllocatorsWithSlotsToBackpatch_Locked()`
  - At the end of a `LoaderAllocator`'s lifetime, the `LoaderAllocator` is unregistered from dependency `LoaderAllocators`, see `MethodDescBackpatchInfoTracker::ClearDependencyMethodDescEntryPointSlotsToBackpatchHash()`
  - When a `MethodDesc`'s entry point changes, backpatching also includes iterating over recorded dependent `LoaderAllocators` to backpatch the relevant slots recorded there, see `BackpatchEntryPointSlots()`

Synchronization between entry point changes and backpatching slots
  - A global lock is used to ensure that all recorded backpatchable slots corresponding to a `MethodDesc` point to the same entry point, see `DoBackpatch()` and `BackpatchEntryPointSlots()` for examples

Typical slot value transitions:
  - Initially, the slot contains the method's temporary entry point, which always points to the prestub (see above)
  - After the tier 0 JIT completes, the slot is transitioned to the tier 0 entry point, and the slot is recorded for backpatching
  - When tiered compilation decides to begin counting calls for the method, the slot is transitioned to the temporary entry point (call counting currently happens in the prestub)
  - When the call count reaches the tier 1 threshold, the slot is transitioned to the tier 0 entry point and a tier 1 JIT is scheduled
  - After the tier 1 JIT completes, the slot is transitioned to the tier 1 entry point

Due to startup time perf issues:
  - `IsEligibleForTieredCompilation()` is called more frequently with this change and in hotter paths. I chose to use a `MethodDesc` flag to store that information for fast retreival. The flag is initialized by `DetermineAndSetIsEligibleForTieredCompilation()`.
  - Initially, I experimented with allowing a tiered vtable method to have a precode, and allocated a new precode that would also be the stable entry point when a direct call is necessary. That also allows recording a new slot to be optional - in the event of an OOM, the slot may just point to the stable entry point. There are a large number of such methods and the allocations were slowing down startup perf. So, I had to eliminate precodes for tiered vtable methods and that in turn means that recording slots is necessary for versionability.

@jkotas @davidwrighton @noahfalk @AndyAyersMS

kouvel commented:
For perf, here is what I have so far. I'm in the process of gathering data for some more tests, and more samples for some cases to weed out error, I'll update inline once I have more data. Startup perf is a bit lower in some JitBench tests (measurably but slightly), and same as before in TechEmpower ASP.NET tests. Steady-state perf is better in several TechEmpower ASP.NET tests.

For memory, I took a count of number of methods eligible for vtable slot backpatching and in JitBench tests it's approximately 1/3 of methods eligible for tiering. Typically, there are no slots recorded for a method eligible for vtable slot backpatching. In the same tests, cases where there is a slot recorded, there is typically only one or two. From that high level info I figured the memory impact would be minimal and scales proportionally with number of virtual methods. I can gather more specific data if you would like, let me know.

JitBench - Before

```
       Benchmark                Metric           NoR2R NoTiering           NoR2R               Default
-----------------------  --------------------  -------------------  -------------------  -------------------
Dotnet_Build_HelloWorld         Duration (ms)     2906 (2893-2931)     1972 (1970-1989)   1193.5 (1192-1199)
        Csc_Hello_World         Duration (ms)     1893 (1880-1897)     1002 (1001-1023)        463 (462-467)
      Csc_Roslyn_Source         Duration (ms)     3656 (3637-3716)     3267 (3246-3281)     2381 (2356-2393)
  Empty Console Program         Duration (ms)         41.5 (40-44)           40 (40-41)           41 (40-43)
             MusicStore          Startup (ms)     1721 (1717-1744)     1061 (1059-1067)        524 (516-529)
             MusicStore    First Request (ms)     1666 (1658-1677)        926 (918-931)        414 (413-416)
             MusicStore  Median Response (ms)    2.025 (2.01-2.04)     2.05 (2.03-2.07)     2.04 (2.02-2.05)
               AllReady          Startup (ms)     2442 (2437-2455)     1692 (1686-1701)     1114 (1108-1119)
               AllReady    First Request (ms)     1018 (1016-1021)        594 (593-603)      339.5 (336-341)
               AllReady  Median Response (ms)      2.57 (2.55-2.6)      2.63 (2.6-2.66)     2.69 (2.65-2.72)
               Word2Vec         Training (ms)  33302 (33136-33683)  33472 (33049-33778)  34886 (34482-35125)
               Word2Vec     First Search (ms)           24 (24-24)           24 (24-24)           24 (24-24)
               Word2Vec    Median Search (ms)  21.52 (21.51-21.71)  21.52 (21.48-21.64)   21.52 (21.5-21.54)
```

JitBench - After

```
       Benchmark                Metric           NoR2R NoTiering           NoR2R               Default
-----------------------  --------------------  -------------------  -------------------  -------------------
Dotnet_Build_HelloWorld         Duration (ms)     2927 (2917-2944)   1992.5 (1989-1994)     1230 (1223-1231)
        Csc_Hello_World         Duration (ms)     1876 (1870-1884)    1003.5 (997-1004)      468.5 (467-473)
      Csc_Roslyn_Source         Duration (ms)     3639 (3608-3700)     3188 (3184-3246)     2323 (2316-2337)
  Empty Console Program         Duration (ms)           42 (40-44)           40 (40-41)           40 (40-41)
             MusicStore          Startup (ms)     1713 (1705-1721)     1074 (1070-1091)        536 (531-538)
             MusicStore    First Request (ms)     1653 (1651-1664)        931 (923-940)        424 (420-427)
             MusicStore  Median Response (ms)        2.02 (2-2.04)     2.04 (2.01-2.06)     2.02 (2.01-2.06)
               AllReady          Startup (ms)     2458 (2441-2467)     1712 (1707-1720)     1131 (1121-1134)
               AllReady    First Request (ms)     1020 (1017-1031)        598 (594-599)      344.5 (342-346)
               AllReady  Median Response (ms)     2.54 (2.53-2.57)     2.54 (2.52-2.58)     2.61 (2.58-2.63)
               Word2Vec         Training (ms)  33831 (33375-34061)  33674 (33153-34666)  34967 (34064-35293)
               Word2Vec     First Search (ms)         24.5 (24-25)           24 (24-24)         24.5 (24-25)
               Word2Vec    Median Search (ms)  21.51 (21.48-21.65)   21.56 (21.5-21.61)   21.8 (21.74-21.86)
```

TechEmpower ASP.NET
- With fragile ngen for CoreLib, R2R for CoreFX libraries, and IL-only for ASP.NET libraries

```
                       Untiered    TieredBefore     TieredAfter
Json Windows KestrelLibuv 3.0
       Averages       445303.50       437595.75       443278.75
         Diff %                           -1.73           -0.45
Json Linux KestrelLibuv 3.0
       Averages       415051.75       405420.25       414559.25
         Diff %                           -2.32           -0.12
MvcDbFortunesEf Windows KestrelSockets 3.0
       Averages        47756.50        46654.50        48197.50
         Diff %                           -2.31            0.92
MvcDbFortunesRaw Linux KestrelSockets 3.0
       Averages        52792.00        52078.75        53140.75
         Diff %                           -1.35            0.66
MvcJil Windows KestrelSockets 3.0
       Averages       153156.75       150891.50       155253.25
         Diff %                           -1.48            1.37
MvcJil Linux KestrelSockets 3.0
       Averages       151144.75       149524.25       154269.00
         Diff %                           -1.07            2.07
MvcJson Linux KestrelSockets 3.0
       Averages       149530.00       147166.50       151099.25
         Diff %                           -1.58            1.05
MvcJson Linux KestrelLibuv 3.0
       Averages       152232.25       146555.50       149910.00
         Diff %                           -3.73           -1.53
MvcJson Windows KestrelSockets 3.0
       Averages       151553.50       149452.50       153041.75
         Diff %                           -1.39            0.98
MvcPlaintext Windows KestrelSockets 3.0
       Averages       470748.00       462994.50       487390.50
         Diff %                           -1.65            3.54
MvcPlaintext Linux KestrelSockets 3.0
       Averages       440622.00       423773.00       444478.75
         Diff %                           -3.82            0.88
MvcPlaintext Linux KestrelLibuv 3.0
       Averages       368242.25       355953.25       378091.25
         Diff %                           -3.34            2.67
Plaintext Linux KestrelSockets 3.0
       Averages      1627354.00      1568912.50      1616499.50
         Diff %                           -3.59           -0.67
ResponseCachingPlaintextCached Linux KestrelSockets 3.0
       Averages       711943.00       676588.75       711204.00
         Diff %                           -4.97           -0.10
ResponseCachingPlaintextCachedDelete Windows KestrelSockets 3.0
       Averages      1772164.00      1742629.50      1794387.25
         Diff %                           -1.67            1.25
ResponseCachingPlaintextVaryByCached Linux KestrelSockets 3.0
       Averages       610905.75       592023.00       620111.00
         Diff %                           -3.09            1.51
StaticFiles Windows KestrelSockets 3.0
       Averages      1643332.25      1540940.25      1653089.25
         Diff %                           -6.23            0.59
StaticFiles Linux KestrelSockets 3.0
       Averages      1417072.50      1376715.00      1437011.50
         Diff %                           -2.85            1.41
```

TechEmpower ASP.NET
- With fragile ngen for CoreLib, no R2R

```
                       Untiered    TieredBefore     TieredAfter
Json Windows KestrelLibuv 3.0
       Averages       462032.75       441298.50       458722.50
         Diff %                           -4.49           -0.72
Json Linux KestrelLibuv 3.0
       Averages       430198.25       411239.50       421508.75
         Diff %                           -4.41           -2.02
MvcJil Windows KestrelSockets 3.0
       Averages       162481.50       159152.00       162845.50
         Diff %                           -2.05            0.22
MvcJil Linux KestrelSockets 3.0
       Averages       159984.25       155254.25       158751.00
         Diff %                           -2.96           -0.77
MvcJson Windows KestrelSockets 3.0
       Averages       163000.25       156076.00       160886.50
         Diff %                           -4.25           -1.30
MvcJson Linux KestrelSockets 3.0
       Averages       154869.75       151463.00       155782.50
         Diff %                           -2.20            0.59
MvcJson Windows KestrelLibuv 3.0
       Averages       148772.25       144894.00       146338.25
         Diff %                           -2.61           -1.64
MvcJson Linux KestrelLibuv 3.0
       Averages       158103.50       150926.50       155319.00
         Diff %                           -4.54           -1.76
MvcPlaintext Linux KestrelLibuv 3.0
       Averages       410122.75       378099.50       400040.25
         Diff %                           -7.81           -2.46
MvcPlaintext Windows KestrelLibuv 3.0
       Averages       418138.75       360367.00       400923.75
         Diff %                          -13.82           -4.12
MvcPlaintext Linux KestrelSockets 3.0
       Averages       472403.25       450942.50       468460.00
         Diff %                           -4.54           -0.83
MvcPlaintext Windows KestrelSockets 3.0
       Averages       518424.75       493111.50       528631.75
         Diff %                           -4.88            1.97
Plaintext Windows KestrelSockets 3.0
       Averages      1935945.25      1884463.25      1934683.75
         Diff %                           -2.66           -0.07
Plaintext Linux KestrelSockets 3.0
       Averages      1678457.00      1632260.00      1663187.50
         Diff %                           -2.75           -0.91
ResponseCachingPlaintextCached Windows KestrelSockets 3.0
       Averages       799342.00       787967.50       816876.25
         Diff %                           -1.42            2.19
ResponseCachingPlaintextCached Linux KestrelSockets 3.0
       Averages       691771.00       712360.75       744267.25
         Diff %                            2.98            7.59
ResponseCachingPlaintextCachedDelete Windows KestrelSockets 3.0
       Averages      1794292.25      1720943.75      1795011.50
         Diff %                           -4.09            0.04
ResponseCachingPlaintextCachedDelete Linux KestrelSockets 3.0
       Averages      1595171.75      1543908.75      1591782.75
         Diff %                           -3.21           -0.21
ResponseCachingPlaintextVaryByCached Windows KestrelSockets 3.0
       Averages       681665.50       670866.25       699504.00
         Diff %                           -1.58            2.62
ResponseCachingPlaintextVaryByCached Linux KestrelSockets 3.0
       Averages       587342.50       623477.75       638033.25
         Diff %                            6.15            8.63
StaticFiles Windows KestrelSockets 3.0
       Averages      1695651.25      1652075.25      1683988.75
         Diff %                           -2.57           -0.69
StaticFiles Linux KestrelSockets 3.0
       Averages      1479025.75      1440269.00      1468742.25
         Diff %                           -2.62           -0.70
```

kouvel commented:
Most of the slight startup time degradation I believe is from finding and dealing with `MethodDesc`s when copying slots to derived types. From using the `MethodData` cache much of that was decreased but `GetMethodDescFromSlot` and `GetMethodDescFromSlotAddress` still appear to take more time during startup than before.

kouvel commented:
Updated commit description

kouvel commented:
Some concrete data on memory impact is below. I have not accounted for unused space in arrays/hashtables and with the numbers below it looks like the amount of unused space would be small. 

- VirtualInfo count = number of MethodDescVirtualInfo objects created
- SlotsToBackpatch count = number of slots recorded for backpatching in the same LoaderAllocator as the MethodDesc
- Dependent LoaderAllocator count = number of dependent LoaderAllocators that have slots to backpatch and are recorded in MethodDescVirtualInfo
- LoaderAllocator dependency MethodDesc count = number of MethodDescs in a different LoaderAllocator for which slots have been recorded in this LoaderAllocator (element count of new hash table in loaderallocator.hpp)
- Dependent SlotsToBackpatch count = number of slots recorded in a LoaderAllocator that is different from the corresponding MethodDesc's LoaderAllocator
- ctvm = called tiered vtable method

Number of slots recorded per called tiered vtable method is low. The MethodDescVirtualInfo itself appears to be more significant by byte count than the number of slots recorded in it. There may be possibilities to tune for 1 and 2 slots without as much overhead. From below though, the overhead looks fairly low to me.

JitBench MusicStore
```
Tiered method count:                               72129
Tiered vtable method count:                        21039
Called tiered method count:                        13068
Called tiered vtable method count:                  3620
VirtualInfo count:                                  1674 (  40 B each,      66960 B,  18.50 B/ctvm)
SlotsToBackpatch count:                             4084 (   8 B each,      32672 B,   9.03 B/ctvm,  1.13 slots/ctvm)
Dependent LoaderAllocator count:                       5 (   8 B each,         40 B,   0.01 B/ctvm)
LoaderAllocator dependency MethodDesc count:           5 (  32 B each,        160 B,   0.04 B/ctvm)
Dependent SlotsToBackpatch count:                      5 (   8 B each,         40 B,   0.01 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  99872 B,  27.59 B/ctvm,  1.13 slots/ctvm)
```

JitBench AllReady
```
Tiered method count:                               75584
Tiered vtable method count:                        21308
Called tiered method count:                        12639
Called tiered vtable method count:                  3458
VirtualInfo count:                                  1821 (  40 B each,      72840 B,  21.06 B/ctvm)
SlotsToBackpatch count:                             7214 (   8 B each,      57712 B,  16.69 B/ctvm,  2.09 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                 130552 B,  37.75 B/ctvm,  2.09 slots/ctvm)
```

JitBench Word2Vec
```
Tiered method count:                                2573
Tiered vtable method count:                          326
Called tiered method count:                          259
Called tiered vtable method count:                    24
VirtualInfo count:                                     1 (  40 B each,         40 B,   1.67 B/ctvm)
SlotsToBackpatch count:                                1 (   8 B each,          8 B,   0.33 B/ctvm,  0.04 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                     48 B,   2.00 B/ctvm,  0.04 slots/ctvm)
```

JitBench CscBuildRoslynSource (repeated in same process many times)
```
Tiered method count:                              112785
Tiered vtable method count:                        34319
Called tiered method count:                        16749
Called tiered vtable method count:                  5346
VirtualInfo count:                                  2014 (  40 B each,      80560 B,  15.07 B/ctvm)
SlotsToBackpatch count:                             7062 (   8 B each,      56496 B,  10.57 B/ctvm,  1.32 slots/ctvm)
Dependent LoaderAllocator count:                       7 (   8 B each,         56 B,   0.01 B/ctvm)
LoaderAllocator dependency MethodDesc count:           7 (  32 B each,        224 B,   0.04 B/ctvm)
Dependent SlotsToBackpatch count:                      9 (   8 B each,         72 B,   0.01 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                 137408 B,  25.70 B/ctvm,  1.32 slots/ctvm)
```

JitBench CscBuildHelloWorld (repeated in same process many times)
```
Tiered method count:                               83313
Tiered vtable method count:                        22174
Called tiered method count:                         6726
Called tiered vtable method count:                  1552
VirtualInfo count:                                   728 (  40 B each,      29120 B,  18.76 B/ctvm)
SlotsToBackpatch count:                             1407 (   8 B each,      11256 B,   7.25 B/ctvm,  0.91 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  40376 B,  26.02 B/ctvm,  0.91 slots/ctvm)
```

AspNetCore Json KestrelLibuv
```
Tiered method count:                               23646
Tiered vtable method count:                         6323
Called tiered method count:                         3657
Called tiered vtable method count:                   589
VirtualInfo count:                                   293 (  40 B each,      11720 B,  19.90 B/ctvm)
SlotsToBackpatch count:                              485 (   8 B each,       3880 B,   6.59 B/ctvm,  0.82 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  15600 B,  26.49 B/ctvm,  0.82 slots/ctvm)
```

AspNetCore MvcJil  KestrelSockets
```
Tiered method count:                               38132
Tiered vtable method count:                         9688
Called tiered method count:                         6211
Called tiered vtable method count:                   944
VirtualInfo count:                                   482 (  40 B each,      19280 B,  20.42 B/ctvm)
SlotsToBackpatch count:                              947 (   8 B each,       7576 B,   8.03 B/ctvm,  1.00 slots/ctvm)
Dependent LoaderAllocator count:                       2 (   8 B each,         16 B,   0.02 B/ctvm)
LoaderAllocator dependency MethodDesc count:           2 (  32 B each,         64 B,   0.07 B/ctvm)
Dependent SlotsToBackpatch count:                      2 (   8 B each,         16 B,   0.02 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  26952 B,  28.55 B/ctvm,  1.01 slots/ctvm)
```

AspNetCore MvcJson KestrelSockets
```
Tiered method count:                               32286
Tiered vtable method count:                         8559
Called tiered method count:                         5761
Called tiered vtable method count:                   932
VirtualInfo count:                                   471 (  40 B each,      18840 B,  20.21 B/ctvm)
SlotsToBackpatch count:                              937 (   8 B each,       7496 B,   8.04 B/ctvm,  1.01 slots/ctvm)
Dependent LoaderAllocator count:                       2 (   8 B each,         16 B,   0.02 B/ctvm)
LoaderAllocator dependency MethodDesc count:           2 (  32 B each,         64 B,   0.07 B/ctvm)
Dependent SlotsToBackpatch count:                      2 (   8 B each,         16 B,   0.02 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  26432 B,  28.36 B/ctvm,  1.01 slots/ctvm)
```

AspNetCore MvcJson KestrelLibuv
```
Tiered method count:                               32080
Tiered vtable method count:                         8598
Called tiered method count:                         5865
Called tiered vtable method count:                   946
VirtualInfo count:                                   473 (  40 B each,      18920 B,  20.00 B/ctvm)
SlotsToBackpatch count:                              940 (   8 B each,       7520 B,   7.95 B/ctvm,  0.99 slots/ctvm)
Dependent LoaderAllocator count:                       2 (   8 B each,         16 B,   0.02 B/ctvm)
LoaderAllocator dependency MethodDesc count:           2 (  32 B each,         64 B,   0.07 B/ctvm)
Dependent SlotsToBackpatch count:                      2 (   8 B each,         16 B,   0.02 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  26536 B,  28.05 B/ctvm,  1.00 slots/ctvm)
```

AspNetCore MvcPlaintext KestrelSockets
```
Tiered method count:                               28795
Tiered vtable method count:                         7526
Called tiered method count:                         5108
Called tiered vtable method count:                   815
VirtualInfo count:                                   427 (  40 B each,      17080 B,  20.96 B/ctvm)
SlotsToBackpatch count:                              842 (   8 B each,       6736 B,   8.27 B/ctvm,  1.03 slots/ctvm)
Dependent LoaderAllocator count:                       2 (   8 B each,         16 B,   0.02 B/ctvm)
LoaderAllocator dependency MethodDesc count:           2 (  32 B each,         64 B,   0.08 B/ctvm)
Dependent SlotsToBackpatch count:                      2 (   8 B each,         16 B,   0.02 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  23912 B,  29.34 B/ctvm,  1.04 slots/ctvm)
```

AspNetCore MvcPlaintext KestrelLibuv
```
Tiered method count:                               28589
Tiered vtable method count:                         7565
Called tiered method count:                         5212
Called tiered vtable method count:                   829
VirtualInfo count:                                   429 (  40 B each,      17160 B,  20.70 B/ctvm)
SlotsToBackpatch count:                              841 (   8 B each,       6728 B,   8.12 B/ctvm,  1.01 slots/ctvm)
Dependent LoaderAllocator count:                       2 (   8 B each,         16 B,   0.02 B/ctvm)
LoaderAllocator dependency MethodDesc count:           2 (  32 B each,         64 B,   0.08 B/ctvm)
Dependent SlotsToBackpatch count:                      2 (   8 B each,         16 B,   0.02 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  23984 B,  28.93 B/ctvm,  1.02 slots/ctvm)
```

AspNetCore Plaintext KestrelSockets
```
Tiered method count:                               20092
Tiered vtable method count:                         5081
Called tiered method count:                         3051
Called tiered vtable method count:                   490
VirtualInfo count:                                   258 (  40 B each,      10320 B,  21.06 B/ctvm)
SlotsToBackpatch count:                              403 (   8 B each,       3224 B,   6.58 B/ctvm,  0.82 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  13544 B,  27.64 B/ctvm,  0.82 slots/ctvm)
```

AspNetCore ResponseCachingPlaintextCached KestrelSockets
```
Tiered method count:                               21323
Tiered vtable method count:                         5283
Called tiered method count:                         3429
Called tiered vtable method count:                   570
VirtualInfo count:                                   297 (  40 B each,      11880 B,  20.84 B/ctvm)
SlotsToBackpatch count:                              483 (   8 B each,       3864 B,   6.78 B/ctvm,  0.85 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  15744 B,  27.62 B/ctvm,  0.85 slots/ctvm)
```

AspNetCore ResponseCachingPlaintextCachedDelete KestrelSockets
```
Tiered method count:                               21716
Tiered vtable method count:                         5518
Called tiered method count:                         3099
Called tiered vtable method count:                   480
VirtualInfo count:                                   253 (  40 B each,      10120 B,  21.08 B/ctvm)
SlotsToBackpatch count:                              403 (   8 B each,       3224 B,   6.72 B/ctvm,  0.84 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  13344 B,  27.80 B/ctvm,  0.84 slots/ctvm)
```

AspNetCore ResponseCachingPlaintextVaryByCached KestrelSockets
```
Tiered method count:                               21335
Tiered vtable method count:                         5284
Called tiered method count:                         3459
Called tiered vtable method count:                   573
VirtualInfo count:                                   301 (  40 B each,      12040 B,  21.01 B/ctvm)
SlotsToBackpatch count:                              469 (   8 B each,       3752 B,   6.55 B/ctvm,  0.82 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  15792 B,  27.56 B/ctvm,  0.82 slots/ctvm)
```

AspNetCore StaticFiles KestrelSockets
```
Tiered method count:                               20309
Tiered vtable method count:                         5091
Called tiered method count:                         3110
Called tiered vtable method count:                   497
VirtualInfo count:                                   262 (  40 B each,      10480 B,  21.09 B/ctvm)
SlotsToBackpatch count:                              419 (   8 B each,       3352 B,   6.74 B/ctvm,  0.84 slots/ctvm)
Dependent LoaderAllocator count:                       0 (   8 B each,          0 B,   0.00 B/ctvm)
LoaderAllocator dependency MethodDesc count:           0 (  32 B each,          0 B,   0.00 B/ctvm)
Dependent SlotsToBackpatch count:                      0 (   8 B each,          0 B,   0.00 B/ctvm,  0.00 slots/ctvm)
Total:                                                   (                  13832 B,  27.83 B/ctvm,  0.84 slots/ctvm)
```

davidwrighton commented:
>     // Keep in-sync with MethodDesc::IsEligibleForTieredCompilation()

Change this to MethodDesc::DetermineAndSetIsEligibleForTieredCompilation

---
Refers to: src/vm/methodtablebuilder.cpp:6952 in c294ed4. [](commit_id = c294ed47a75c70d49b991a17adebffca1979b304, deletion_comment = False)

davidwrighton commented on src/vm/methodtablebuilder.cpp:
```diff
@@ -8832,10 +8832,13 @@ void MethodTableBuilder::CopyExactParentSlots(MethodTable *pMT, MethodTable *pAp
+                MethodDesc *pMD = hCanonMTData->GetImplMethodDesc(i);
+                CONSISTENCY_CHECK(CheckPointer(pMD));
+                CONSISTENCY_CHECK(pMD == pCanonMT->GetMethodDescForSlot(i));
+                pMT->SetSlot(i, pMD->GetInitialEntryPointForCopiedSlot());
```
Please break this out into a helper function. I see it copied in at least 4 places in the codebase.

kouvel commented:
Updated perf numbers and added count of new FuncPtrStubs to memory numbers inline above. Rebased for conflicts. Addressed feedback in the last two commits.

davidwrighton commented:
Why is all this infrastructure labeled with Virtual this and VTable that? I don't see any fundamental reason why this stuff should be tied to virtual functions. It seems to me that this would also be useful for other places where we hold a function pointer, that ideally would just point to the target method address. For instance, in R2R code all calls to other methods go through an indirection cell, and this concept of a backpatchable address could be used to improve the performance of those calls for applications which are relatively short lived.

Also, I'm in the process of building out infrastructure to support making interface dispatch calls to interfaces which only have 1 implementation. This would involve placing code pointers directly into VSD indirection cells. Interestingly enough, this exactly matches the logic here, except that those cells need to be able to be updated to not be target address pointers. It feels like that would be a straightforward addition to this logic, but I'd love to hear your opinion.

davidwrighton commented:
Finally, I've been looking at this code for a few days, and I've concluded that the memory management is correct, and to the best of my ability to code review the implementation is correct, but I'm uneasy with the way that cross loader allocator data is handled.

1. The data structure is implemented in a such a way that a similar data structure would require rebuilding it all from scratch, while necessary for this purpose, it feels like a data structure that should be written in a generic fashion.
2. The cost of a LoaderAllocatorSet per MethodDesc that goes multi-loaderallocator seems somewhat high in terms of memory consumption. The numbers around memory use presented above are not representative of the true cost here, as you are not really measuring any collectible loaderallocator scenarios, as well as the minimum cost of having a LoaderAllocatorSet which is ~32 bytes + 7*pointersize. I have been thinking that instead of attempting to use a LoaderAllocatorSet, we could play some games with a hashtable and seperate storage for the nodes, such that we store the nodes in the appropriate loaderallocator, but a hashtable of pointers in the loaderallocator of the controlling method. This would reduce lookup costs to a single O(~1) lookup, reduce memory usage by removing the need for duplicate LoaderAllocatorSets, and be roughly equivalent in cleanup costs as the cleanup of the one master hash is probably not any more expensive than cleaning an item out of a large collection of small sets.

Overall, I believe the current state is acceptable for checkin, but I would like to see the data structure management turn into a generic service of the LoaderAllocator concept in the next few months.

kouvel commented:
Thanks for the feedback. You're right, the actual storage doesn't need to be specific to virtualness. I'll go ahead and remove association with virtualness for the storage portion.

As discussed offline, the LoaderAllocatorSet can be eliminated in favor of using a multi-hash instead and that could simplify some of the data structures and save some space. I'll look into how easy it would be to make that change as well, so that it would be easy to move to the generic service in the future.

kouvel commented:
Rebased, last three commits are new. After discussing further offline, I kept the storage mechanism as it was, and did some cleanup to make it easier to move to a new storage mechanism in the future and to use the same backpatching logic for other purposes.

kouvel commented:
@dotnet-bot test this please

kouvel commented:
@davidwrighton, would you be able to make a pass over the latest commits (starting from the one titled "Rename VirtualInfo -> BackpatchInfo")? I'd like to get this into 3.0 preview 2 (I'm told that would be around mid Jan) with some bake time to flush things out.

kouvel commented:
@davidwrighton I see you're on vacation, this can wait, have a good time!

noahfalk commented on tests/src/baseservices/TieredCompilation/TieredVtableMethodTests.cs:
```diff
@@ -0,0 +1,235 @@
+        for (int i = 0; i < 4; ++i)
+        {
+            CallVirtualMethod(baseObj, CallCountPerIteration);
+            CallVirtualMethod(derivedObj, CallCountPerIteration);
```
I suggest adding a virtual method with a generic method parameter and call it with both valuetype and reference type arguments.

kouvel commented on src/vm/frames.cpp:
```diff
@@ -735,13 +734,12 @@ BOOL StubDispatchFrame::TraceFrame(Thread *thread, BOOL fromPatch,
     if (fromPatch)
     {
-        trace->InitForStub(GetFunction()->GetStableEntryPoint());
+        trace->InitForStub(GetFunction()->GetMethodEntryPoint());
```
Thanks for the details, makes sense now, done

kouvel commented on src/vm/precode.cpp:
```diff
@@ -521,7 +545,7 @@ TADDR Precode::AllocateTemporaryEntryPoints(MethodDescChunk *  pChunk,
     // Note that these are just best guesses to save memory. If we guessed wrong,
     // we will allocate a new exact type of precode in GetOrCreatePrecode.
-    BOOL fForcedPrecode = pFirstMD->RequiresStableEntryPoint(count > 1);
+    BOOL fForcedPrecode = hasMethodDescVersionableWithPrecode || pFirstMD->RequiresStableEntryPoint(count > 1);
```
The call to `RequiresStableEntryPoint(true /* estimateForChunk */)` would estimate for the chunk, but the estimation would not necessarily correct for the versionability check. This is the only place where `true` is passed for the argument, the code here tries to compensate for that. For untiered methods the estimate doesn't have to be correct but tiered methods require a backpatchable precode.

kouvel replied:
Actually the estimation doesn't have to be correct anyway, a wrong estimation would just cause additional precodes to be allocated instead of sharing one backpatchable precode along with the temporary entry point. Probably doesn't matter much.

kouvel commented on src/vm/precode.cpp:
```diff
@@ -494,6 +494,30 @@ TADDR Precode::AllocateTemporaryEntryPoints(MethodDescChunk *  pChunk,
+        MethodDesc *pMD = pChunk->GetFirstMethodDesc();
+        for (int i = 0; i < count; ++i)
+        {
+            if (pMD->DetermineAndSetIsEligibleForTieredCompilation())
```
I didn't see a significant difference when it's called once per method. I think you're right though, could consider improving that in the future.

kouvel commented on Documentation/design-docs/code-versioning.md:
````diff
@@ -332,7 +332,8 @@ In order to do step 3 the CodeVersionManager relies on one of two different mech
 
 ```
-MethodDesc::IsVersionableWithPrecode()
+MethodDesc::IsEligibleForTieredCompilation() && MethodDesc::IsTieredMethodVersionableWithPrecode()
````
Updated to make vtable slot backpatching a function of code versioning rather than a function of tiering. Restored the other `IsVersionableWith*` functions and added an `IsVersionableWithoutJumpStamp()`, which I think would be useful in the future as well with possibly more versioning techniques.

kouvel commented on src/vm/appdomain.cpp:
```diff
@@ -3904,6 +3904,12 @@ void AppDomain::Terminate()
+    // Recorded entry point slots may point into the virtual call stub manager's heaps, so clear it first
+    GetLoaderAllocator()
+        ->GetMethodDescBackpatchInfoTracker()
+        ->ClearDependencyMethodDescEntryPointSlotsToBackpatchHash(GetLoaderAllocator());
```
I didn't want to pollute the namespace with names that are too generic, and wanted the names to be a bit more precise than above. I made an attempt to simplify them a bit.

kouvel commented on src/vm/frames.cpp:
```diff
@@ -568,13 +568,12 @@ BOOL PrestubMethodFrame::TraceFrame(Thread *thread, BOOL fromPatch,
     if (fromPatch)
     {
-        trace->InitForStub(GetFunction()->GetStableEntryPoint());
+        trace->InitForStub(GetFunction()->GetMethodEntryPoint());
```
Thanks for the info, makes sense conceptually

kouvel commented on src/vm/method.cpp:
```diff
@@ -2188,6 +2188,12 @@ PCODE MethodDesc::TryGetMultiCallableAddrOfCode(CORINFO_ACCESS_FLAGS accessFlags
     if (HasStableEntryPoint())
         return GetStableEntryPoint();
 
+    if (IsTieredVtableMethod())
```
The fallthrough path doesn't always return null. It may return the temporary entry point and for these methods that precode always points to the prestub. This was the only call to `MayHavePrecode()` and I had elected to assert instead of changing it, I have changed it now.

kouvel commented on src/vm/method.cpp:
```diff
@@ -4798,6 +4811,214 @@ Precode* MethodDesc::GetOrCreatePrecode()
+
+    MethodDescBackpatchInfo *backpatchInfo =
+        mdLoaderAllocator->GetMethodDescBackpatchInfoTracker()->GetOrAddBackpatchInfo_Locked(this);
+    if (slotLoaderAllocator == mdLoaderAllocator)
```
Currently for recorded slots, the slot's LoaderAllocator is either the same as the MethodDesc's, or it's a LoaderAllocator dependent on the MethodDesc, so the slot's LoaderAllocator's lifetime will be <= the MethodDesc's LoaderAllocator's lifetime. The MethodDesc can't disappear while there are still slots referring to its code. I'm not sure when it would be a common case.

kouvel replied:
If A and B have AppDomain lifetime, why would they not use the same LoaderAllocator? Maybe they don't currently but perhaps that could be a better solution?

noahfalk replied:
I might be mistaken and maybe they already do, but I thought it was one per module.

kouvel commented on src/vm/method.cpp:
```diff
@@ -4798,6 +4811,214 @@ Precode* MethodDesc::GetOrCreatePrecode()
+    SetEntryPointToBackpatch_Locked(entryPoint, isPrestubEntryPoint);
+}
+
+void MethodDesc::SetCodeEntryPoint(PCODE entryPoint)
```
I didn't see a good way to fully avoid ambiguity while keeping the names simple, I also think there is already some ambiguity in the names. There are SetMethodEntryPoint(), SetStableEntryPointInterlocked(), SetNativeCodeInterlocked(), SetCodeEntryPoint(), PublishNativeCode().

In my view SetMethod... and SetStable... refer to the initial entry point that will be called. That may be the code entry point, it may be a precode that redirects to the code entry point, it may be a some other stub, etc.

SetCode... refers to the actual code entry point that the initial entry point will eventually redirect to. When the method's entry point is the code entry point, SetCode... would update the method's entry point. When the method's entry point is a precode, SetCode... would update the precode's target. SetCode... also handles backpatching other slots that also point directly to the method's code entry point.

SetNativeCode... seems to be an odd mix. If there is a native code slot, it just stores the code entry point in the slot without updating any actual entry point, but the stored code entry point may be used by callers. The prestub may later update the method's entry point with the stored code entry point. Tiering uses the native code slot as a storage location for tier 0 code, so SetNativeCode... cannot be called for other code versions. If there is no native code slot, SetNativeCode... does something similar to SetCode... but for more specific cases.

Of the lot, I'm actually most satisfied with the name SetCodeEntryPoint(). I think I'll leave it as is for now, maybe something can be done in the future to rename or refactor these functions to decrease ambiguity and simplify/clarify their purpose.

kouvel replied:
It is possible to define the terms a bit differently and less restrictively while still keeping true to the sense of the terms.

EntryPoint could be defined as anything that can be called by a caller using an appropriate ABI for that EntryPoint and MethodDesc to execute the method correctly. For the purposes of discussion at least, an indirection cell could also be considered an EntryPoint if it contains a value that is an EntryPoint, it is just an indirect EntryPoint. Under that definition the results of some cases above may be different:
- The start of jitted or prejitted code for a method typically is an EntryPoint
- A Precode is an EntryPoint
- The target of a Precode may be an EntryPoint
- A stub may or may not be an EntryPoint depending on its ABI requirements. For example, a Prestub is not an EntryPoint, but a CallCountingStub for tiering could be an EntryPoint.
- For code versionable with a JumpStamp, the start of jitted or prejitted code for any version is an EntryPoint. Though it is preferable that the first version is used in multi-callable cases to maintain versionability, for a single call it may be correct to call a different version directly.

In SetCodeEntryPoint(), what I meant by Code is the start of any code that immediately begins executing the method's code.
- The start of jitted or prejitted code for the method is a CodeEntryPoint
- A Precode may or may not be a CodeEntryPoint. If the target of the Precode is a CodeEntryPoint, then the Precode is not a CodeEntryPoint. If on the other hand, the ABI requires calls to go through the Precode, then the Precode is a CodeEntryPoint.
- The target of a Precode may be a CodeEntryPoint
- An entry point into the interpreter that starts executing method's IL is a CodeEntryPoint
- A CallCountingStub could be considered a CodeEntryPoint if it is considered that it is just an implementation detail that a stub is used. Call counting may be considered part of the method's code as a whole, just with that portion separated into a stub.

NativeCode could be used for cases that specifically refer to compiled code. A MethodDesc could be agnostic about the type of code that a code slot contains. For instance, it may be useful to store an interpreter entry point in the NativeCodeSlot but it currently would be a violation of the name.

I'll go ahead and leave it as is for now. Once we figure out what we want the names to be, the names can be cleaned up in a separate PR.

noahfalk replied:
Sounds good to me. Thanks Kount!

kouvel commented on src/vm/method.hpp:
```diff
@@ -1202,28 +1204,191 @@ class MethodDesc
+        return IsEligibleForTieredCompilation() && IsTieredMethodVersionableWithVtableSlotBackpatch();
+    }
+
+    bool IsEligibleForEntryPointSlotBackpatch()
```
I have renamed it to MayHaveEntryPointSlotsToBackpatch(). The backpatching data structures can be used for a more general purpose than for backpatching vtable slots. So IsVersionableWithVtableSlotBackpatch() as I called it, refers to the versioning scheme specific to vtable slots, which would have specific restrictions, and MayHaveEntryPointSlotsToBackpatch() in the future would be true for any versioning scheme or method that has a slot recorded for backpatching. I figure in the future there may be another versioning scheme that handles slots in caller code for backpatching direct calls, or one that handles non-vtable slots, and there may also be other uses for it.

kouvel commented on src/vm/methoddescbackpatchinfo.cpp:
```diff
@@ -0,0 +1,241 @@
+    }
+
+    NewHolder<MethodDescEntryPointSlotsToBackpatch> methodDescSlotsHolder =
+        new MethodDescEntryPointSlotsToBackpatch(methodDesc);
```
I think it's better to throw. For an app that intends to work well in memory-restricted environments, it may well be OOM-tolerant. If it is not tolerant then it's likely to fail pretty soon anyway so it doesn't matter much. Throwing would allow a retry to succeed without penalty and I'd prefer to avoid locking into a slow system with no recovery unless necessary. Going through the prestub on every call can be slow enough that it could appear as a bug and not just a perf issue. If the prestub could attempt to record slots that failed to be recorded before, that could be a solution, but not all of the info to do that is available currently and it may not be sustainable longer-term.

kouvel replied:
A partial solution may be to catch the OOM from recording a slot in DoBackpatch() and let it try again on the next call to the prestub, and let the OOM propagate for VSD cases.

kouvel commented on src/vm/prestub.cpp:
```diff
@@ -83,58 +82,113 @@ PCODE MethodDesc::DoBackpatch(MethodTable * pMT, MethodTable *pDispatchingMT, BO
 
-    PCODE pExpected = GetTemporaryEntryPoint();
+        // True interface methods are never backpatched and do not qualify for tiered vtable slot backpatching
+        _ASSERTE(!(pMT->IsInterface() && !IsStatic()));
```
I'm not sure why it would have been there before, but when I tried to enable it I saw an issue related to default interface methods and method impls. It looked like backpatching a method impl slot on an interface type can be invalid. IIRC it was the [diamondshape test](https://github.com/dotnet/coreclr/tree/master/tests/src/Loader/classloader/DefaultInterfaceMethods/diamondshape) that failed due to calling a wrong method as a result of incorrect backpatching but I don't remember the details now.

kouvel commented on src/vm/amd64/cgenamd64.cpp:
```diff
@@ -880,10 +880,10 @@ EXTERN_C PCODE VirtualMethodFixupWorker(TransitionBlock * pTransitionBlock, CORC
         INSTALL_UNWIND_AND_CONTINUE_HANDLER_NO_PROBE;
 
-        if (pMD->IsTieredVtableMethod())
+        if (pMD->IsVersionableWithVtableSlotBackpatch())
```
Yea I went back and forth on that. The VSD slots that are recorded are sort of extensions of vtable slots, in the sense they copy the value of a vtable slot and so they must be backpatched whenever the source vtable slot changes. They are not actually vtable slots but slots that cache vtable slot values, I thought the naming was close enough. Another reason to include "vtable" or something similar in the name is because the versioning scheme is specific to methods that have a vtable slot, which would have different restrictions (like not having a precode or stable entry point for example) than another versioning scheme that may apply to non-vtable slots or caller slots.

kouvel commented:
Rebased for conflicts

kouvel commented:
@dotnet-bot test Tizen armel Cross Checked Innerloop Build and Test

kouvel commented:
Tizen armel seems to be some infra issue that is unrelated to this change:
```
16:14:39 ERROR: Unable to find project for artifact copy: dotnet_corefx/master/tizen_armel_cross_release
16:14:39 This may be due to incorrect project name or permission settings; see help for project name in job configuration.
```
